File: projects/ios-readplace/App/ReaderBootstrap.swift
import Foundation

enum ReaderSessionMint: Equatable {
	case minted([HTTPCookie])
	case failed
	case superseded
}

enum ReaderBootstrap: Equatable {
	case loading
	case ready([HTTPCookie])
	case unavailable

	init(after mint: ReaderSessionMint) {
		switch mint {
		case .minted(let cookies):
			self = .ready(cookies)
		case .failed:
			self = .unavailable
		case .superseded:
			self = .loading
		}
	}
}

File: projects/native-apps/android/app/src/main/kotlin/com/readplace/android/app/ReaderBootstrap.kt
package com.readplace.android.app

import okhttp3.Cookie

sealed interface ReaderSessionMint {
	data class Minted(val cookies: List<Cookie>) : ReaderSessionMint

	data object Failed : ReaderSessionMint

	data object Superseded : ReaderSessionMint
}

sealed interface ReaderBootstrap {
	data object Loading : ReaderBootstrap

	data class Ready(val cookies: List<Cookie>) : ReaderBootstrap

	data object Unavailable : ReaderBootstrap

	companion object {
		fun after(mint: ReaderSessionMint): ReaderBootstrap =
			when (mint) {
				is ReaderSessionMint.Minted -> Ready(mint.cookies)
				ReaderSessionMint.Failed -> Unavailable
				ReaderSessionMint.Superseded -> Loading
			}
	}
}
